custom_envs.py

```diff
@@ -0,0 +1,111 @@
+from gym_minigrid.envs.crossing import LavaCrossingEnv
+import numpy as np
+from queue import Queue
+
+class LavaCrossingSpotRewardEnv(LavaCrossingEnv):
+  def __init__(self):
+    super().__init__()
+    self.training = True
+    self.reward_grid = None
+    self.goal_pos = (0, 0)
+    self.reset()
+
+  def _gen_reward(self):
+    # generate the reward values in the grid (use the numpy array reward_grid)
+    # first, get shortest path to goal for every square in the grid, then take
+    # the inverse of the values to get our P function (the higher it is, the
+    # closer we are to the goal
+    bfs_q = Queue()
+    processed = np.zeros_like(self.reward_grid)
+
+    # append array positions, neighbor's distance to goal to the queue, we will
+    # process a square by setting its proper distance value, adding unprocessed
+    # neighbors, modifying the # processed mask, and popping the current
+    # position from the queue
+    bfs_q.put((self.goal_pos, 0))
+    while not bfs_q.empty():
+      # parse queue item
+      pos, neighbor_dist = bfs_q.get()
+      r, c = pos
+      neighbors = [(r, c+1), (r+1, c), (r, c-1), (r-1, c)]
+
+      # if it is the goal, we don't change the reward value, otherwise, the
+      # shortest path is 1 + neighbor's distance
+      if self.reward_grid[r, c] != 2:
+        self.reward_grid[r, c] = neighbor_dist + 1
+
+      for r_n, c_n in neighbors:
+        # skip positions that are out of bounds
+        if r_n >= self.reward_grid.shape[0]: continue
+        elif c_n >= self.reward_grid.shape[1]: continue
+        # don't process unreachable squares (lava or wall, marked as 1
+        elif self.reward_grid[r_n, c_n] == 1: continue
+        # don't process squares that have already been processed
+        elif processed[r_n, c_n]: continue
+
+        # add the neighbor to the queue with the current square's distance
+        bfs_q.put(((r_n, c_n), self.reward_grid[r, c]))
+
+        # set the current neighbor as processed
+        processed[r_n, c_n] = 1
+
+    # finally, invert the values in the reward grid to get P
+    # use 2 in the numerator so that reaching the goal has reward of 1
+    self.reward_grid = 2 / self.reward_grid
```

Comment:
I don't think this will be a linear increase in reward as you progress towards the goal. I guess the progress definition doesn't technically require that, so let's leave it as-is, and we can change the code only if it proves necessary.

Reply:
Yeah, we could modify to be max_dist - self.reward_grid to maintain linearity if necessary